cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -189,6 +191,72 @@ Status ConstBoolExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
   return Status::OK();
 }
 
+struct NonZeroVisitor {
+   UInt32Builder *builder;
+   const ArrayData& array;
+
+   NonZeroVisitor(UInt32Builder *builder, const ArrayData& array)
+   : builder(builder), array(array) {}
+
+  Status Visit(const DataType& type) {
+    return Status::TypeError("Unsupported type for nonzero: ", type.ToString());
+  }
+
+   template <typename Type> 
+   enable_if_t<has_c_type<Type>::value && 
+              !std::is_same<Type, MonthDayNanoIntervalType>::value &&
+              !std::is_same<Type, DayTimeIntervalType>::value, 
+              Status>
+   Visit(const Type&) {
+     using T = typename GetViewType<Type>::T;
+     uint32_t index = 0;
+
+     return VisitArrayDataInline<Type>(
+        this->array,
+        [&](T v) {
+          if(v != 0) {
```

Comment:
Since `T` can be `bool` (besides integers and floating-point), maybe using `if (v) { ... }` is a more general expression.

Reply:
:+1: